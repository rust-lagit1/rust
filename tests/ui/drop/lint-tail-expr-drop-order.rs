// Edition 2024 lint for change in drop order at tail expression
// This lint is to capture potential change in program semantics
// due to implementation of RFC 3606 <https://github.com/rust-lang/rfcs/pull/3606>

#![deny(tail_expr_drop_order)]
#![feature(shorter_tail_lifetimes)]

struct LoudDropper;
impl Drop for LoudDropper {
    fn drop(&mut self) {
        // This destructor should be considered significant because it is a custom destructor
        // and we will assume that the destructor can generate side effects arbitrarily so that
        // a change in drop order is visible.
        println!("loud drop");
    }
}
impl LoudDropper {
    fn get(&self) -> i32 {
        0
    }
}

fn should_lint() -> i32 {
    let x = LoudDropper;
    // Should lint
    x.get() + LoudDropper.get()
    //~^ ERROR: this value of type `LoudDropper` has significant drop implementation that will have a different drop order from that of Edition 2021
    //~| WARN: this changes meaning in Rust 2024
}

fn should_not_lint_closure() -> impl FnOnce() -> i32 {
    let x = LoudDropper;
    move || {
        // Should not lint
        x.get() + LoudDropper.get()
    }
}

fn should_lint_in_nested_items() {
    fn should_lint_me() -> i32 {
        let x = LoudDropper;
        // Should lint
        x.get() + LoudDropper.get()
        //~^ ERROR: this value of type `LoudDropper` has significant drop implementation that will have a different drop order from that of Edition 2021
        //~| WARN: this changes meaning in Rust 2024
    }
}

fn should_lint_params(x: LoudDropper) -> i32 {
    x.get() + LoudDropper.get()
    //~^ ERROR: this value of type `LoudDropper` has significant drop implementation that will have a different drop order from that of Edition 2021
    //~| WARN: this changes meaning in Rust 2024
}

fn should_not_lint() -> i32 {
    let x = LoudDropper;
    // Should not lint
    x.get()
}

fn should_lint_in_nested_block() -> i32 {
    let x = LoudDropper;
    { LoudDropper.get() }
    //~^ ERROR: this value of type `LoudDropper` has significant drop implementation that will have a different drop order from that of Edition 2021
    //~| WARN: this changes meaning in Rust 2024
}

fn should_not_lint_in_match_arm() -> i32 {
    let x = LoudDropper;
    // Should not lint because Edition 2021 drops temporaries in blocks earlier already
    match &x {
        _ => LoudDropper.get(),
    }
}

fn should_not_lint_when_consumed() -> (LoudDropper, i32) {
    let x = LoudDropper;
    // Should not lint
    (LoudDropper, x.get())
}

struct MyAdt {
    a: LoudDropper,
    b: LoudDropper,
}

fn should_not_lint_when_consumed_in_ctor() -> MyAdt {
    let a = LoudDropper;
    // Should not lint
    MyAdt { a, b: LoudDropper }
}

fn should_not_lint_when_moved() -> i32 {
    let x = LoudDropper;
    drop(x);
    // Should not lint
    LoudDropper.get()
}

fn main() {}
