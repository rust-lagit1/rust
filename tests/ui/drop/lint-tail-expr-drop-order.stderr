error: this value of type `LoudDropper` has significant drop implementation that will have a different drop order from that of Edition 2021
  --> $DIR/lint-tail-expr-drop-order.rs:26:15
   |
LL |     let x = LoudDropper;
   |         - these local bindings with significant drop implementation may observe changes in drop order under Edition 2024
LL |     // Should lint
LL |     x.get() + LoudDropper.get()
   |               ^^^^^^^^^^^
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see issue #123739 <https://github.com/rust-lang/rust/issues/123739>
note: the lint level is defined here
  --> $DIR/lint-tail-expr-drop-order.rs:5:9
   |
LL | #![deny(tail_expr_drop_order)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: this value of type `LoudDropper` has significant drop implementation that will have a different drop order from that of Edition 2021
  --> $DIR/lint-tail-expr-drop-order.rs:43:19
   |
LL |         let x = LoudDropper;
   |             - these local bindings with significant drop implementation may observe changes in drop order under Edition 2024
LL |         // Should lint
LL |         x.get() + LoudDropper.get()
   |                   ^^^^^^^^^^^
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see issue #123739 <https://github.com/rust-lang/rust/issues/123739>

error: this value of type `LoudDropper` has significant drop implementation that will have a different drop order from that of Edition 2021
  --> $DIR/lint-tail-expr-drop-order.rs:50:15
   |
LL | fn should_lint_params(x: LoudDropper) -> i32 {
   |                       - these local bindings with significant drop implementation may observe changes in drop order under Edition 2024
LL |     x.get() + LoudDropper.get()
   |               ^^^^^^^^^^^
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see issue #123739 <https://github.com/rust-lang/rust/issues/123739>

error: this value of type `LoudDropper` has significant drop implementation that will have a different drop order from that of Edition 2021
  --> $DIR/lint-tail-expr-drop-order.rs:63:7
   |
LL |     let x = LoudDropper;
   |         - these local bindings with significant drop implementation may observe changes in drop order under Edition 2024
LL |     { LoudDropper.get() }
   |       ^^^^^^^^^^^
   |
   = warning: this changes meaning in Rust 2024
   = note: for more information, see issue #123739 <https://github.com/rust-lang/rust/issues/123739>

error: aborting due to 4 previous errors

